File: android/projeto-leilao/app/src/main/java/br/com/alura/leilao/asynctask/BaseAsyncTask.java
package br.com.alura.leilao.asynctask;

import android.os.AsyncTask;
import android.support.annotation.NonNull;

/**
 * @author Cristian Urbainski
 * @since 1.0 (20/09/20)
 */
public class BaseAsyncTask<T> extends AsyncTask<Void, Void, T> {

    private final ActionListener<T> actionListener;
    private ResultListener<T> resultListener;

    public BaseAsyncTask(@NonNull ActionListener<T> actionListener, ResultListener<T> resultListener) {

        this.actionListener = actionListener;
        this.resultListener = resultListener;
    }

    @Override
    protected T doInBackground(Void... voids) {

        return actionListener.execute();
    }

    @Override
    protected void onPostExecute(T t) {

        super.onPostExecute(t);

        if (resultListener != null) {

            resultListener.result(t);
        }
    }

    public interface ActionListener<T> {

        T execute();

    }

    public interface ResultListener<T> {

        void result(T result);

    }
}

File: android/tech-news-app/app/src/main/java/br/com/alura/technews/asynctask/BaseAsyncTask.kt
package br.com.alura.technews.asynctask

import android.os.AsyncTask

class BaseAsyncTask<T>(
    private val quandoExecuta: () -> T,
    private val quandoFinaliza: ((resultado: T) -> Unit)? = null
) : AsyncTask<Void, Void, T>() {

    override fun doInBackground(vararg params: Void?) = quandoExecuta()

    override fun onPostExecute(resultado: T) {
        super.onPostExecute(resultado)

        if (quandoFinaliza != null) {
            quandoFinaliza?.invoke(resultado)
        }
    }

}